File: plugins/plugin-linear/typescript/src/providers/activity.ts
import type { IAgentRuntime, Memory, Provider, State } from "@elizaos/core";
import type { LinearService } from "../services/linear";
import type { LinearActivityItem } from "../types";

export const linearActivityProvider: Provider = {
  name: "LINEAR_ACTIVITY",
  description: "Provides context about recent Linear activity",
  get: async (runtime: IAgentRuntime, _message: Memory, _state: State) => {
    try {
      const linearService = runtime.getService<LinearService>("linear");
      if (!linearService) {
        return {
          text: "Linear service is not available",
        };
      }

      const activity = linearService.getActivityLog(10);

      if (activity.length === 0) {
        return {
          text: "No recent Linear activity",
        };
      }

      const activityList = activity.map((item: LinearActivityItem) => {
        const status = item.success ? "✓" : "✗";
        const time = new Date(item.timestamp).toLocaleTimeString();
        return `${status} ${time}: ${item.action} ${item.resource_type} ${item.resource_id}`;
      });

      const text = `Recent Linear Activity:\n${activityList.join("\n")}`;

      return {
        text,
        data: {
          activity: activity.slice(0, 10).map((item) => ({
            id: item.id,
            action: item.action,
            resource_type: item.resource_type,
            resource_id: item.resource_id,
            success: item.success,
            error: item.error,
            details: JSON.stringify(item.details) as string,
            timestamp:
              typeof item.timestamp === "string"
                ? item.timestamp
                : new Date(item.timestamp).toISOString(),
          })) as Array<Record<string, string | boolean | undefined>>,
        },
      };
    } catch (_error) {
      return {
        text: "Error retrieving Linear activity",
      };
    }
  },
};

File: plugins/plugin-linear/python/src/elizaos_plugin_linear/providers/activity.py
from datetime import datetime

from elizaos_plugin_linear.providers.base import Provider, ProviderResult, RuntimeProtocol
from elizaos_plugin_linear.services.linear import LinearService


async def get_activity(
    runtime: RuntimeProtocol,
    _message: object,
    _state: object,
) -> ProviderResult:
    try:
        linear_service: LinearService = runtime.get_service("linear")
        if not linear_service:
            return ProviderResult(text="Linear service is not available")

        activity = linear_service.get_activity_log(10)

        if not activity:
            return ProviderResult(text="No recent Linear activity")

        activity_list = []
        for item in activity:
            status = "✓" if item.success else "✗"
            try:
                time_str = datetime.fromisoformat(item.timestamp.replace("Z", "")).strftime("%H:%M")
            except Exception:
                time_str = "?"

            activity_list.append(
                f"{status} {time_str}: {item.action} {item.resource_type} {item.resource_id}"
            )

        text = "Recent Linear Activity:\n" + "\n".join(activity_list)

        return ProviderResult(
            text=text,
            data={
                "activity": [
                    {
                        "id": item.id,
                        "action": item.action,
                        "resource_type": item.resource_type,
                        "success": item.success,
                    }
                    for item in activity[:10]
                ]
            },
        )
    except Exception:
        return ProviderResult(text="Error retrieving Linear activity")


linear_activity_provider = Provider(
    name="LINEAR_ACTIVITY",
    description="Provides context about recent Linear activity",
    get=get_activity,
)
